### PR TITLE
STY: Bump pyright, pyupgrade and mypy for new PEP-696 syntax

### PR DESCRIPTION
**Pull request by mroeschke**
Also bumps flake8 and pre-commit which should have innocuous changes

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -1006,7 +1006,7 @@ def _intersection(self, other, sort):
             # duplicates
             taken = self._intersection_non_unique(other)
 
-        if sort is None:
+        if sort:
```
Comment (rhshadrach) [inline pandas/core/indexes/interval.py:1009]: This looks like an actual behavior change? Unless we never actually hit here with a truthy `sort`.

File: pandas/core/arrays/boolean.py [hunk 2]
```diff
@@ -369,7 +369,7 @@ def _coerce_to_array(
             assert dtype == "boolean"
         return coerce_to_array(value, copy=copy)
 
-    def _logical_method(self, other, op):
+    def _logical_method(self, other, op):  # type: ignore[override]
```
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/arrays/boolean.py:372]: Here's the mypy error

```bash
pandas/core/arrays/boolean.py:372: error: Signature of "_logical_method" incompatible with supertype "BaseMaskedArray"  [override]
pandas/core/arrays/boolean.py:372: note:      Superclass:
pandas/core/arrays/boolean.py:372: note:          def _arith_method(self: BaseMaskedArray, other: Any, op: Any) -> Any
pandas/core/arrays/boolean.py:372: note:      Subclass:
pandas/core/arrays/boolean.py:372: note:          def _logical_method(self, other: Any, op: Any) -> Any
pandas/core/arrays/boolean.py:372: note:      Superclass:
pandas/core/arrays/boolean.py:372: note:          def _arith_method(self: BaseMaskedArray, other: Any, op: Any) -> Any
pandas/core/arrays/boolean.py:372: note:      Subclass:
pandas/core/arrays/boolean.py:372: note:          def _logical_method(self, other: Any, op: Any) -> Any
```

In BasedMaskedArray we assign `_logical_method = _arith_method`, but override `_logical_method` in BooleanArray. It looks mypy is doing some inference on `self` even though we don't type it?

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/interval.py:1009]: AFAICT based on the tests and the type annotations this should not be `None`, only `: bool`. Not sure if that's totally intentional though...

Comment (rhshadrach) [reply pandas/core/indexes/interval.py:1009]: This goes back to #42293. On the base implementation, `sort=None` was handled:

https://github.com/pandas-dev/pandas/blob/17ecb56c219d8d01399faffcffa8f059ecf6fd45/pandas/core/indexes/base.py#L6587

Now `sort=None` is still allowed, but it seems to expect only True/False:

https://github.com/pandas-dev/pandas/blob/97c4ce390ae7dafb7f58a21184cc32d921edb6c6/pandas/core/indexes/base.py#L7659`

From this, I think it's just some residual `None` logic that needs a bit more cleaning.

File: pandas/core/arrays/boolean.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/arrays/boolean.py:372]: I think `self` is typed implicitly, yes. Makes sense that it's an override issue, thanks.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Pull request merged**